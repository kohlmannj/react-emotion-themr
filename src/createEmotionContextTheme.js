export function createEmotionContextTheme(contextThemeObj) {
  if (typeof contextThemeObj !== 'object' || contextThemeObj === null) {
    return contextThemeObj;
  }

  return Object.keys(contextThemeObj).reduce(
    (emotionContextTheme, componentName) => ({
      ...emotionContextTheme,
      [componentName]: createEmotionCssModule(contextThemeObj[componentName]),
    }),
    {}
  );
}
