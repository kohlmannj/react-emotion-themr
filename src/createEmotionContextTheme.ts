import { TReactCSSThemrTheme } from '@friendsofreactjs/react-css-themr';
import {
  createEmotionTheme,
  TReactEmotionThemrTheme,
  RenderedEmotionTheme,
} from './createEmotionTheme';

export type TReactEmotionThemrContextTheme = {
  [key: string]: TReactEmotionThemrTheme;
};

export function createEmotionContextTheme<T>(
  contextThemeObj: T
): RenderedEmotionTheme<T, TReactEmotionThemrContextTheme> {
  if (
    typeof contextThemeObj === 'object' &&
    typeof contextThemeObj !== 'undefined' &&
    contextThemeObj !== null
  ) {
    return Object.keys(contextThemeObj).reduce(
      (emotionContextTheme, componentName) => ({
        ...emotionContextTheme,
        [componentName]: createEmotionTheme(
          (contextThemeObj as TReactEmotionThemrTheme)[componentName]
        ),
      }),
      {} as TReactCSSThemrTheme
    ) as T extends TReactEmotionThemrContextTheme ? TReactCSSThemrTheme : never;
  }

  return contextThemeObj as T extends TReactEmotionThemrContextTheme ? never : T;
}
