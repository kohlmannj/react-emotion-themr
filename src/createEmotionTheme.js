import { css } from 'emotion';

export function createEmotionTheme(classNamesObj) {
  if (typeof classNamesObj !== 'object' || classNamesObj === null) {
    return classNamesObj;
  }

  return Object.keys(classNamesObj).reduce((emotionTheme, className) => {
    const value = classNamesObj[className];
    const typeofValueIsNonNullObject = typeof value === 'object' && value !== null;

    // Only add the className if `value` is a non-null object or a string
    if (typeofValueIsNonNullObject || typeof value === 'string') {
      return {
        ...emotionTheme,
        [className]: typeofValueIsNonNullObject ? css(value, `label:${className};`) : value,
      };
    }

    return emotionTheme;
  }, {});
}
