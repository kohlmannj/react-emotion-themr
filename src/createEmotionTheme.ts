import { TReactCSSThemrTheme } from '@friendsofreactjs/react-css-themr';
import { CSSObject } from 'create-emotion';
import { css } from 'emotion';

export type TReactEmotionThemrTheme = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: CSSObject | any;
};

export type RenderedEmotionTheme<T, V = TReactEmotionThemrTheme> = T extends V
  ? TReactCSSThemrTheme
  : T;

export function createEmotionTheme<T>(classNamesObj: T): RenderedEmotionTheme<T> {
  if (
    typeof classNamesObj === 'object' &&
    typeof classNamesObj !== 'undefined' &&
    classNamesObj !== null
  ) {
    return Object.keys(classNamesObj).reduce(
      (emotionTheme, className) => {
        const value = (classNamesObj as TReactEmotionThemrTheme)[className];

        if ((typeof value === 'object' && value !== null) || typeof value === 'string') {
          return {
            ...emotionTheme,
            [className]:
              typeof value === 'object' && value !== null
                ? css(value, `label:${className};`)
                : value,
          } as TReactCSSThemrTheme;
        }

        return emotionTheme;
      },
      {} as TReactCSSThemrTheme
      /** @see https://github.com/microsoft/TypeScript/issues/22735#issuecomment-376960435 */
    ) as T extends TReactEmotionThemrTheme ? TReactCSSThemrTheme : never;
  }

  return classNamesObj as T extends TReactEmotionThemrTheme ? never : T;
}
