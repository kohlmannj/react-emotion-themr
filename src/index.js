export * from './createEmotionContextTheme';
export * from './createEmotionTheme';
export * from './ThemeProvider';
export * from './themr';
