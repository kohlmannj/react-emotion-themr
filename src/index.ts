export * from './createEmotionContextTheme';
export * from './createEmotionTheme';
export * from './ThemeProvider';
export * from './themeable';
export * from './themr';
