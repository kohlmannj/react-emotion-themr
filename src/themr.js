import React from 'react';
import { themr as cssThemr } from '@friendsofreactjs/react-css-themr';
import { createEmotionTheme } from './createEmotionTheme';

/**
 * Maps props and theme to an object that will be used to pass down props to the
 * decorated component. Based on `defaultMapThemrProps()` from @friendsofreactjs/react-css-themr:
 * @see https://github.com/FriendsOfReactJS/react-css-themr/blob/3.3.0/src/components/themr.js#L312-L334
 *
 * @param {Object} ownProps - All props given to the decorated component
 * @param {Object} theme - Calculated then that should be passed down
 * @returns {Object} - Props that will be passed down to the decorated component
 */
function mapThemrPropsWithEmotion(ownProps, theme) {
  const {
    composeTheme, //eslint-disable-line no-unused-vars
    innerRef,
    themeNamespace, //eslint-disable-line no-unused-vars
    mapThemrProps, //eslint-disable-line no-unused-vars
    ...rest
  } = ownProps;

  return {
    ...rest,
    ref: innerRef,
    theme: createEmotionTheme(theme),
  };
}

export function themr(componentName, localTheme, { mapThemrProps, ...restOptions } = {}) {
  return ThemedComponent =>
    cssThemr(componentName, createEmotionTheme(localTheme), {
      ...restOptions,
      mapThemrProps: mapThemrProps
        ? // For a custom `mapThemrProps()` function, make sure we transform `theme` first
          (ownProps, theme) => mapThemrProps(ownProps, createEmotionTheme(theme))
        : // By default, transform `theme` into `createEmotionTheme(theme)` (see above)
          mapThemrPropsWithEmotion,
    })(ThemedComponent);
}
