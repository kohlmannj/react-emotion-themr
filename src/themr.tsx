import React, { ComponentType } from 'react';
import {
  themr as cssThemr,
  IThemrOptions,
  TMapThemrProps,
} from '@friendsofreactjs/react-css-themr';
import { createEmotionTheme } from './createEmotionTheme';
import { TReactEmotionThemrTheme } from './createEmotionTheme';

export function themr(
  identifier: string | number | symbol,
  defaultTheme?: TReactEmotionThemrTheme,
  options?: IThemrOptions
) {
  return function emotionThemr<P = {}>(
    component: ComponentType<P>
  ): ComponentType<P & { mapThemrProps?: TMapThemrProps<P>; theme?: TReactEmotionThemrTheme }> {
    // Decorate the ThemedComponent
    const CssThemed = cssThemr(identifier, createEmotionTheme(defaultTheme), options)(component);

    // Wrap CssThemed in another component which knows how to handle the `theme` prop
    const EmotionThemed = (
      props: P & { mapThemrProps?: TMapThemrProps<P>; theme?: TReactEmotionThemrTheme }
    ): JSX.Element => {
      return (
        // eslint-disable-next-line compat/compat
        <CssThemed {...Object.assign({}, props, { theme: createEmotionTheme(props.theme) })} />
      );
    };

    EmotionThemed.displayName = `EmotionThemed${component.displayName || 'Component'}`;

    return EmotionThemed;
  };
}
