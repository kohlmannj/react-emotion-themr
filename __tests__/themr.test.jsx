import React from 'react';
import { shallow } from 'enzyme';
import { themr } from '../src/themr';

const BaseComponent = ({ className, theme, ...rest }) => (
  <div className={[className, theme.root].join(' ')} {...rest} />
);

describe('themr', () => {
  it('creates a decorated React component', () => {
    const MyThemedComponent = themr('MyThemedComponent')(BaseComponent);
    const wrapper = shallow(<MyThemedComponent />).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('creates a decorated React component that can accept `theme` props with CSS property-value objects', () => {
    const MyThemedComponent = themr('MyThemedComponent')(BaseComponent);
    const wrapper = shallow(
      <MyThemedComponent theme={{ root: { backgroundColor: 'rebeccapurple' } }} />
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
