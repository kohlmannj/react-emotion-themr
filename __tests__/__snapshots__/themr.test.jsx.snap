// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`themr creates a decorated React component 1`] = `
<BaseComponent
  theme={Object {}}
/>
`;

exports[`themr creates a decorated React component that can accept \`theme\` props with CSS property-value objects 1`] = `
<BaseComponent
  theme={
    Object {
      "root": "css-ux9s12-root",
    }
  }
/>
`;
